File: utilities/numbers/go/number_utils.go
package numbers

import (
	"math"
	"strconv"
)

func Clamp(value, minimum, maximum float64) float64 {
	return math.Min(math.Max(value, minimum), maximum)
}

func IsNumeric(value string) bool {
	if value == "" { return false }
	_, err := strconv.ParseFloat(value, 64)
	return err == nil
}

func MinValue(values []float64) *float64 {
	if len(values) == 0 { return nil }
	result := values[0]
	for _, value := range values[1:] {
		if value < result { result = value }
	}
	return &result
}

func MaxValue(values []float64) *float64 {
	if len(values) == 0 { return nil }
	result := values[0]
	for _, value := range values[1:] {
		if value > result { result = value }
	}
	return &result
}

File: utilities/numbers/csharp/NumberUtils.cs
namespace PolyglotUtilities.Numbers;

public static class NumberUtils
{
    public static double Clamp(double value, double minimum, double maximum)
    {
        return Math.Min(Math.Max(value, minimum), maximum);
    }

    public static bool IsNumeric(string? value)
    {
        return value is not null && value.Length > 0 && double.TryParse(
            value,
            System.Globalization.NumberStyles.Float,
            System.Globalization.CultureInfo.InvariantCulture,
            out _);
    }

    public static double? MinValue(IEnumerable<double> values)
    {
        var materialized = values.ToList();
        return materialized.Count == 0 ? null : materialized.Min();
    }

    public static double? MaxValue(IEnumerable<double> values)
    {
        var materialized = values.ToList();
        return materialized.Count == 0 ? null : materialized.Max();
    }
}
